src/versionbits.cpp

```diff
@@ -49,20 +56,19 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
         pindexPrev = vToCompute.back();
         vToCompute.pop_back();
 
+        // We track state by previous-block, so the height we should be comparing is +1
+        const int64_t height = pindexPrev->nHeight + 1;
+
         switch (state) {
             case ThresholdState::DEFINED: {
-                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
+                if (height >= height_timeout) {
                     stateNext = ThresholdState::FAILED;
```

Comment:
This doesn't match the BIP 8 spec which specifies that DEFINED always transitions to STARTED and never skips directly to FAILED.

Reply:
This condition was actually caught by the fuzzer (back when this PR included it).

The scenario is that the timeoutheight is less than one period after the startheight. Without this, such a deployment would transition to the STARTED state, and then potentially to LOCKED_IN. But all of that after the timeoutheight has been passed. That doesn't make any sense and doesn't fit within what we are expecting in any tests.

I think we need to revisit this case in the BIP.

Reply:
After discussing this with @luke-jr, I've removed this transition. Since there should never be a deployment that has the timeoutheight less than one period after the startheight, I've also added a sanity check to the unit tests for this case.

Reply:
Yes, the "always reach STARTED" issue was why there's a separate NEVER_ACTIVE setting, I think.

BIP8 does specify "timeoutheight must be at least 4096 blocks (2 retarget intervals) after startheight", so you could treat timeoutheight being reached from DEFINED as an extension to the spec and define it any way you like, similar to the always/never active cases. (The old bip8 code in #19573 didn't do that in order to match the spec in the bip as closely as possible to catch spec/implementation mismatches)

As far as the fuzz tester goes, following the bip8 spec exactly should mean something like:

``c++
switch (state) {
case STARTED:
    assert(startheight <= height);
    if (exp_state == STARTED) {
        assert(height <= timeoutheight);
        assert(blocks_sig < threshold);
    } else {
        assert(exp_state == DEFINED);
    }
    break;
case FAILED:
    assert(exp_state == STARTED || exp_state == FAILED);
    assert(never_active_test || height >= timeoutheight);
} 
```   

or alternatively extending it so you can skip STARTED would probably not change as much but might warrant adding a check along the lines of:

```c++
static int next_boundary(int n, int m) { return (n + m - 1) - ((n - 1) % m); }
if (next_boundary(timeoutheight, period) >= next_boundary(startheight, period)) {
    assert(state == DEFINED || state == FAILED);
    assert(exp_state == DEFINED || exp_state == FAILED);
}
```